flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartition.java

```diff
@@ -83,24 +86,45 @@
 	/** The total number of bytes (both data and event buffers). */
 	private long totalNumberOfBytes;
 
+	/** The collection of buffers which are spanned over by checkpoint barrier and needs to be persisted for snapshot. */
+	private final List<Buffer> inflightBufferSnapshot = new ArrayList<>();
+
 	// ------------------------------------------------------------------------
 
 	PipelinedSubpartition(int index, ResultPartition parent) {
 		super(index, parent);
 	}
 
 	@Override
-	public boolean add(BufferConsumer bufferConsumer) {
-		return add(bufferConsumer, false);
+	public boolean add(BufferConsumer bufferConsumer) throws IOException {
+		return add(bufferConsumer, false, false);
 	}
 
 	@Override
 	public void finish() throws IOException {
-		add(EventSerializer.toBufferConsumer(EndOfPartitionEvent.INSTANCE), true);
+		add(EventSerializer.toBufferConsumer(EndOfPartitionEvent.INSTANCE), true, false);
 		LOG.debug("{}: Finished {}.", parent.getOwningTaskName(), this);
 	}
 
-	private boolean add(BufferConsumer bufferConsumer, boolean finish) {
+	@Override
+	public EnqueueAction priorityEvent(AbstractEvent event, BufferConsumer eventBufferConsumer) throws IOException {
+		if (readView != null) {
+			switch (readView.priorityEvent(event, eventBufferConsumer)) {
```

Comment:
I think we can guarantee to insert the event in head for any cases. Refer to the comments in `LocalBufferPool` method. Then wen can remove the return value from `#priorityEvent`.

Reply:
I think for the `PipelinedSubpartition` and `BoundedBlockingSubpartition` we do not need to return those values.  We could just define `void addPriorityEvent(...)` method in the `ResultSubpartition` method instead of inheriting from `PriorityEventListener`. The same applies for `PipelinedSubpartitionView`, as those classes are not being used as listeners, they are just forwarding the call. This extra abundant inheritance makes the code a little bit less readable and messes up with IDE support of usages.

Reply:
One more thing. @AHeise shouldn't we merge `BufferAvailabilityListener` and `PriorityEventListener` interfaces? They are doing basically the same thing. One is notifying about "hot" data, another about normal data and they should be used together always (especially after dropping `PriorityEventListener` inheritance from `ResultSubpartition` and `PipelinedSubpartitionView`).

Reply:
Merging `PriorityEventListener` into `BufferAvailabilityListener` also greatly helped to reduce commit size.